Fixed remove items from stash on create market offer

dudantas opened the pull request:
Resolves #470 

Created the Player::requestLockerItems function to avoid code repetition, it can be used in ProtocolGame and Game class improved the way locker items were read and stored

Game::getMarketItemList function has been replaced by Player::requestLockerItems
Small modification in ProtocolGame::sendMarketEnter function to avoid code repetition, it will read Player::requestLockerItems function

Added some error logs for market functions so that we can identify where the problem was, in future tests. And if there is a problem during the processing of the offer, it will not be created, making it impossible to have an item or money clone.

Added some logs (debug) to make it easier to find problems in the future

dudantas committed:
Fix remove items from stash on create market offer

CleberMoreira reviewed (approved):
Tested with low and very high values like 20k+ items at once and it was smooth! 110% approved.

gerotib commented:
When we have certain item both in depot and stash, and create market offer, it will remove the amount from both depot and stash. It should prioritize either depot or stash, so no double amount of items is being removed.

dudantas commented:
> When we have certain item both in depot and stash, and create market offer, it will remove the amount from both depot and stash. It should prioritize either depot or stash, so no double amount of items is being removed.

Ok I'll check.

dudantas committed:
Fix market offers

dudantas committed:
Merge branch 'main' into fix-remove-stash-item-create-market-offer

dudantas commented:
@gerotib the bugs have been corrected. Or the only thing that was left was related to item count. When I have more than 65k items stored and try to raise an offer, it won't create. I'm investigating.

dudantas committed:
Merge branch 'main' into fix-remove-stash-item-create-market-offer

dudantas committed:
Fix market offer amount

dudantas committed:
Fix functions order

dudantas committed:
Fix compilation

dudantas committed:
New createMarketOffer get items from locker and others changes

Created the Player::requestLockerItems function to avoid code repetition, it can be used in ProtocolGame and Game
improved the way locker items were read and stored

Game::getMarketItemList function has been replaced by Player::requestLockerItems
Small modification in ProtocolGame::sendMarketEnter function to avoid code repetition, it will read Player::requestLockerItems function

Added some logs (debug) to make it easier to find problems in the future

dudantas committed:
Fix compilation

dudantas committed:
Fix variable name

dudantas commented:
@gerotib and @CleberMoreira can test again, please?

dudantas committed:
Remove unecessary check

dudantas committed:
Some fixes

dudantas committed:
Some sonar fixes

dudantas committed:
Merge branch 'main' into fix-remove-stash-item-create-market-offer

dudantas committed:
Fix typo

dudantas committed:
Remove unecessary ItemType

guispiller commented:
Found a small issue there, this is how to reproduce:
1 - put 94k of the same item on the stash (example rope belt)
2 - create a sell offer of 64k rope belts. (you'll still have 30k on stash)
3 - cancel your sell offer of rope belts. (you'll have 64k on mailbox and 30k on stash)
4 - create new 64k sell offer on market. All items will be removed from stash and from mailbox (you lose 30k items)

Also found another bug, but i think it's on main too:
if you press to stow items from mailbox, it's checking and stowing items from player's bp instead of mailbox.

dudantas committed:
Rework on createMarketOffer and acceptMarket offer

Extracted remove items in own function
Added some checks for avoid item cloning/duplication

guispiller commented:
After the last commits, the bugs I commented were solved and also the stow function of mailbox were implemented.

I just noticed that, when you right click on a stackable item inside the mailbox, and select "stow all items of this type", the function is searching items to stow at the players backpack, not at the mailbox.

dudantas committed:
Fix stow inbox and depot items

dudantas commented:
> After the last commits, the bugs I commented were solved and also the stow function of mailbox were implemented.
> 
> I just noticed that, when you right click on a stackable item inside the mailbox, and select "stow all items of this type", the function is searching items to stow at the players backpack, not at the mailbox.

Thanks for reporting the issues. Are fixed!

dudantas committed:
Fix stow items

dudantas committed:
Fix log

gerotib commented:
@dudantas I have found an issue in `removeOfferItems` function, item is not being removed while creating market offer if we have some stackable items in inbox. It takes the count from the stackable ones inside inbox instead of actual item id we try to sell. Its possible to clone items this way. 
https://gfycat.com/smallhighbonobo

dudantas commented:
> @dudantas I have found an issue in `removeOfferItems` function, item is not being removed while creating market offer if we have some stackable items in inbox. It takes the count from the stackable ones inside inbox instead of actual item id we try to sell. Its possible to clone items this way. https://gfycat.com/smallhighbonobo

@gerotib It's been ten days since I asked to test again. Now, I've merged this pr. I'll test it later and let you know. You can open an issue?

gerotib commented:
@dudantas sure, I just got back and didn't see this earlier
